sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -187,8 +188,17 @@ object FileFormatWriter extends Logging {
     // We should first sort by partition columns, then bucket id, and finally sorting columns.
     val requiredOrdering =
       partitionColumns ++ writerBucketSpec.map(_.bucketIdExpression) ++ sortColumns
+
+    // SPARK-40588: plan may contain an AdaptiveSparkPlanExec, which does not know
+    // its final plan's ordering, so we have to materialize that plan first
+    def materializeAdaptiveSparkPlan(plan: SparkPlan): SparkPlan = plan match {
```

Comment:
I am not sure what you mean with "is brough"` in "the missing ordering field of Sort in the top layer is also brought"?

It is also not clear to me what you are referring to with "removing 'AdaptiveSparkPlanExec'". Do you mean the change in this this PR? The 'AdaptiveSparkPlanExec' is not _removed_ but _replaced_ with the final plan. It is the same plan that is otherwise used when calling `.execute`, so AQE should work as before.

Reply:
`V1WriteCommand` is a concept introduced after Spark 3.3. The issue addressed in this PR does not exist post Spark 3.3 as the `FieFormatWriter` does not see `AdaptiveSparkPlanExec` but the final plan (exactly what this change is trying to achieve).